### ISSUE TITLE
Android: Flickering when hardware acceleration is enabled - Fix included

### ISSUE THREAD

**Issue by bailuk** [mapsforge/mapsforge#919]
I open a new issue report because this affects many reported issues and is (hopefully) of important new information.

## Abstract
I found the cause for the supper annoying flickering in the map view when hardware acceleration is enabled. Attached is a replacement for the FrameBuffer class that removes the flickering almost entirely. I have tested this on three different devices (Among them an old and underpowered smart phone with a single core processor and a tablet with a big screen).

The fix is very hackish and there are still some issues with timing (there is only a small time frame where the bitmaps can be swapped).

## Now what did provoke the flickering?
In software rendering the canvas modifies the underlying bitmap immediately. For example:
When canvas.fillColor(..) returns, inside the underlying bitmap (or view) all pixels are set to new color.

In hw accelerated rendering the canvas only collects the drawing commands. The actual execution of this commands (like set all pixel to a specific color) is done at some time later.

Now when LayerManager.doWork was called twice in row and MapView.invalidate() has triggered MapView.onDraw() only once the canvas has still stored a reference to the old bitmap (the reference is from the collected canvas.drawBitmap() command). That old bitmap is now the same bitmap that LayerManager.doWork() is modifying. So what happens in this situation is that the same bitmaps that gets drawn on the screen gets modified  by the LayerManager as well.

## What did I change?
In my version of the FrameBuffer I've made the following changes:
- LayerManager.doWork locks a bitmap before drawing and releases it after drawing.
- The bitmaps are swapped in the draw() function (UI thread) right before canvas.drawBitmap() is called.  
- The bitmaps are only swapped if LayerManager has not locked a bitmap.

```

public class FrameBufferHack extends FrameBuffer {
    private static final boolean IS_TRANSPARENT = false;

    /**
     * odBitmap: onDraw bitmap  -
     *           the bitmap that gets displayed by the MapView.onDraw() function from the UI thread
     * lmBitmap: layerManager bitmap -
     *           the bitmap that gets drawn by the LayerManager.doWork() thread
     */
    private Bitmap odBitmap;
    private Bitmap lmBitmap;
    private MapPosition lmMapPosition;

    private final Object lmBitmapLock = new Object();
    private boolean isLmBitmapLocked = false;
    private boolean haveNewLmBitmap = false;


    private Dimension dimension;
    private final DisplayModel displayModel;
    private final FrameBufferModel frameBufferModel;
    private final GraphicFactory graphicFactory;
    private final Matrix matrix;

    public FrameBufferHack(Model model) {
        super(model.frameBufferModel, model.displayModel, AndroidGraphicFactory.INSTANCE);
        this.frameBufferModel = model.frameBufferModel;
        this.displayModel = model.displayModel;
        this.graphicFactory = AndroidGraphicFactory.INSTANCE;
        this.matrix = graphicFactory.createMatrix();
    }





    // this is called from MapView.onDraw()
    public void draw(GraphicContext graphicContext) {

        graphicContext.fillColor(this.displayModel.getBackgroundColor());

        /**
         * swap bitmaps right before the canvas.drawBitmap function to prevent flickering as much
         * as possible
         */
        swapBitmaps();
        if (this.odBitmap != null) {
            graphicContext.drawBitmap(this.odBitmap, this.matrix);
        }
    }


    private void swapBitmaps() {
        synchronized (lmBitmapLock) {
            /**
             * Swap bitmaps only if the layerManager is currently not working and
             * has drawn a new bitmap since the last swap
             */
            if (! this.isLmBitmapLocked && this.haveNewLmBitmap) {
                Bitmap bitmapTemp = this.odBitmap;
                this.odBitmap = this.lmBitmap;
                this.lmBitmap = bitmapTemp;

                this.haveNewLmBitmap = false;
                this.frameBufferModel.setMapPosition(lmMapPosition);
            }
        }
    }


    // this is called from layer manager when drawing starts
    public Bitmap getDrawingBitmap() {

        // Give the UI thread a chance to swap bitmaps
        if (haveNewLmBitmap) sleep();

        lockLmBitmap();

        if (this.lmBitmap != null) {
            this.lmBitmap.setBackgroundColor(this.displayModel.getBackgroundColor());
        }
        return this.lmBitmap;
    }


    private void lockLmBitmap() {
        synchronized (lmBitmapLock) {
            if (this.lmBitmap != null)
                isLmBitmapLocked = true;
        }
    }

    // this is called from layer manager when drawing is finished
    public void frameFinished(MapPosition frameMapPosition) {
        freeLmBitmap(frameMapPosition);
    }


    private void freeLmBitmap(MapPosition frameMapPosition) {
        synchronized (lmBitmapLock) {
            lmMapPosition = frameMapPosition;
            isLmBitmapLocked = false;
            haveNewLmBitmap = true;
        }
    }




    private static void sleep() {
        try {
            Thread.sleep(10);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public synchronized void adjustMatrix(float diffX, float diffY, float scaleFactor, Dimension mapViewDimension,
                                          float pivotDistanceX, float pivotDistanceY) {
        if (this.dimension == null) {
            return;
        }
        this.matrix.reset();
        centerFrameBufferToMapView(mapViewDimension);
        if (pivotDistanceX == 0 && pivotDistanceY == 0) {
            // only translate the matrix if we are not zooming around a pivot,
            // the translation happens only once the zoom is finished.
            this.matrix.translate(diffX, diffY);
        }

        scale(scaleFactor, pivotDistanceX, pivotDistanceY);
    }

    public synchronized void destroy() {
        destroyBitmaps();
    }






    public synchronized Dimension getDimension() {
        return this.dimension;
    }

    /**
     * @return the bitmap of the second frame to draw on (may be null).
     */


    public synchronized void setDimension(Dimension dimension) {
        if (this.dimension != null && this.dimension.equals(dimension)) {
            return;
        }
        this.dimension = dimension;

        destroyBitmaps();

        if (dimension.width > 0 && dimension.height > 0) {
            this.odBitmap = this.graphicFactory.createBitmap(dimension.width, dimension.height, IS_TRANSPARENT);
            this.lmBitmap = this.graphicFactory.createBitmap(dimension.width, dimension.height, IS_TRANSPARENT);
        }
    }

    private void centerFrameBufferToMapView(Dimension mapViewDimension) {
        float dx = (this.dimension.width - mapViewDimension.width) / -2f;
        float dy = (this.dimension.height - mapViewDimension.height) / -2f;
        this.matrix.translate(dx, dy);
    }

    private void destroyBitmaps() {
        if (this.odBitmap != null) {
            this.odBitmap.decrementRefCount();
            this.odBitmap = null;
        }
        if (this.lmBitmap != null) {
            this.lmBitmap.decrementRefCount();
            this.lmBitmap = null;
        }
    }

    private void scale(float scaleFactor, float pivotDistanceX, float pivotDistanceY) {
        if (scaleFactor != 1) {
            final Point center = this.dimension.getCenter();
            float pivotX = (float) (pivotDistanceX + center.x);
            float pivotY = (float) (pivotDistanceY + center.y);
            this.matrix.scale(scaleFactor, scaleFactor, pivotX, pivotY);
        }
    }

}

```

**Comment by devemux86** [comment]
Thanks for the work and the details!
So after all it seems to be a concurrency issue because of how HA rendering works?
I tested Android & Desktop and not seen anything strange so far, more tests will follow.

Do you want to create a pull request?
Better for start to have the modified `FrameBuffer` as a separate class (like you did).
Let's call it `FrameBufferHA` and I'll add then a static boolean to select which one.
That way users can test the new implementation, without breaking current apps.

Minor remark: in constructor we should use regular `FrameBuffer` parameters, as it's a platform agnostic class, i.e. not tied specifically to Android.

> The fix is very hackish and there are still some issues with timing (there is only a small time frame where the bitmaps can be swapped).

Indeed the sleep call could be improved, as it could cause different behavior depending on device specs.

Related to #613.

**Comment by bailuk** [comment]
> Do you want to create a pull request?

I'd rather leave it to the core developers to write the official implementation.

> So after all it seems to be a concurrency issue because of how HA rendering works?

Exactly!

Here's a slightly modified version of my fix that uses a wait() - notify() mechanism (instead of sleep()) for synchronizing the two bitmap buffers.

 ```
public class FrameBufferHack extends FrameBuffer {
    private static final boolean IS_TRANSPARENT = false;

    /**
     * odBitmap: onDraw bitmap  -
     *           the bitmap that gets displayed by the MapView.onDraw() function from the UI thread
     * lmBitmap: layerManager bitmap -
     *           the bitmap that gets drawn by the LayerManager.doWork() thread
     */


    private Bitmap odBitmap;
    private Bitmap lmBitmap;
    private MapPosition lmMapPosition;

    private final Object lmBitmapLock = new Object();

    private boolean allowBitmapSwap = true;


    private Dimension dimension;
    private final DisplayModel displayModel;
    private final FrameBufferModel frameBufferModel;
    private final GraphicFactory graphicFactory;
    private final Matrix matrix;

    public FrameBufferHack(Model model) {
        super(model.frameBufferModel, model.displayModel, AndroidGraphicFactory.INSTANCE);
        this.frameBufferModel = model.frameBufferModel;
        this.displayModel = model.displayModel;
        this.graphicFactory = AndroidGraphicFactory.INSTANCE;
        this.matrix = graphicFactory.createMatrix();
    }


    // this is called from MapView.onDraw()
    public void draw(GraphicContext graphicContext) {

        graphicContext.fillColor(this.displayModel.getBackgroundColor());

        /**
         * swap bitmaps right before the canvas.drawBitmap function to prevent flickering as much
         * as possible
         */
        swapBitmaps();
        if (this.odBitmap != null) {
            graphicContext.drawBitmap(this.odBitmap, this.matrix);
        }
    }


    private void swapBitmaps() {
        synchronized (lmBitmapLock) {
            /**
             * Swap bitmaps only if the layerManager is currently not working and
             * has drawn a new bitmap since the last swap
             */
            if (allowBitmapSwap) {
                Bitmap bitmapTemp = this.odBitmap;
                this.odBitmap = this.lmBitmap;
                this.lmBitmap = bitmapTemp;


                this.frameBufferModel.setMapPosition(lmMapPosition);

                allowBitmapSwap = false;
                lmBitmapLock.notify();
            }
        }
    }


    // this is called from layer manager when drawing starts
    public Bitmap getDrawingBitmap() {

        lockLmBitmap();


        if (this.lmBitmap != null) {
            this.lmBitmap.setBackgroundColor(this.displayModel.getBackgroundColor());
        }

        return this.lmBitmap;
    }


    private void lockLmBitmap() {
        synchronized (lmBitmapLock) {
            if (this.lmBitmap != null) {
                if (allowBitmapSwap) { // not yet swapped by onDraw()
                    try {
                        lmBitmapLock.wait(); // wait until swapped
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
            allowBitmapSwap = false;
        }
    }

    // this is called from layer manager when drawing is finished
    public void frameFinished(MapPosition frameMapPosition) {
        freeLmBitmap(frameMapPosition);
    }


    private void freeLmBitmap(MapPosition frameMapPosition) {
        synchronized (lmBitmapLock) {
            lmMapPosition = frameMapPosition;
            allowBitmapSwap = true;
        }
    }

[...]

```

**Comment by devemux86** [comment]
I pushed your updated fix in a separate `FrameBufferHA` class (with your name in header).
For now we keep the default `FrameBuffer` class until further tests can be made by all.

I added also [HAMapViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/HAMapViewer.java) in Android samples app, which demonstrates how we can use the new frame buffer. Essentially set before `MapView` creation:
```java
FrameBufferController.HA_FRAME_BUFFER = true;
```
And remember to enable hardware acceleration in Settings.

@bailuk thanks for the work on that!
